lib/blockchain.go

```diff
@@ -517,9 +538,98 @@ type Blockchain struct {
 	syncingState                bool
 	downloadingHistoricalBlocks bool
 
+	// checkpointSyncingProviders is a list of providers from which we will request the committed tip block info
+	// when syncing. The committed tip block info is used to designate a checkpoint before which signature
+	// verification will be skipped. These checkpoint providers should be trusted and should be able to provide
+	// the committed tip block info for the chain we are syncing.
+	checkpointSyncingProviders []string
+	// checkpointBlockInfo is the latest checkpoint block info that we have received from the checkpoint syncing
+	// providers.
+	checkpointBlockInfo *CheckpointBlockInfo
+
 	timer *Timer
 }
 
+func (bc *Blockchain) updateCheckpointBlockInfo() {
+	if len(bc.checkpointSyncingProviders) == 0 {
+		glog.V(2).Info("updateCheckpointBlockInfo: No checkpoint syncing providers set. Skipping update.")
+		return
+	}
+	var wg sync.WaitGroup
+	ch := make(chan *CheckpointBlockInfoAndError, len(bc.checkpointSyncingProviders))
+	for _, provider := range bc.checkpointSyncingProviders {
+		wg.Add(1)
+		go getCheckpointBlockInfoFromProvider(provider, ch, &wg)
+	}
+
+	wg.Wait()
+
+	// Collect the results from the channel
+	checkpointBlockInfos := make([]*CheckpointBlockInfoAndError, len(bc.checkpointSyncingProviders))
+	for ii := range bc.checkpointSyncingProviders {
+		checkpointBlockInfos[ii] = <-ch
+	}
+
+	// Find the checkpoint block info with the highest height.
+	var highestHeightCheckpointBlockInfo *CheckpointBlockInfo
+	for _, checkpointBlockInfo := range checkpointBlockInfos {
+		if checkpointBlockInfo.Error != nil {
+			glog.Errorf("updateCheckpointBlockInfo: Error getting checkpoint block info: %v", checkpointBlockInfo.Error)
+			continue
+		}
+		if highestHeightCheckpointBlockInfo == nil ||
+			checkpointBlockInfo.CheckpointBlockInfo.Height > highestHeightCheckpointBlockInfo.Height {
+			highestHeightCheckpointBlockInfo = checkpointBlockInfo.CheckpointBlockInfo
+		}
+	}
+	if highestHeightCheckpointBlockInfo == nil {
+		glog.Errorf("updateCheckpointBlockInfo: No valid checkpoint block info found.")
+		return
+	}
+	glog.V(2).Infof("updateCheckpointBlockInfo: Setting checkpoint block info to: %v", highestHeightCheckpointBlockInfo)
+	bc.checkpointBlockInfo = highestHeightCheckpointBlockInfo
+}
+
+func getCheckpointBlockInfoFromProvider(provider string, ch chan<- *CheckpointBlockInfoAndError, wg *sync.WaitGroup) {
+	defer wg.Done()
+	url := fmt.Sprintf("%s%s", provider, RoutePathGetCommittedTipBlockInfo)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		ch <- &CheckpointBlockInfoAndError{
```

Comment:
I don't love how easy it is to forget to add to the channel before you return here. I would break it into two functions so you can get compile-time checking on the return like this:

```
func getCheckpointBlockInfoFromProviderHelper(provider string) (*CheckpointBlockInfoAndError) {

}

func getCheckpointBlockInfoFromProvider(provider string, ch chan<- *CheckpointBlockInfoAndError, wg *sync.WaitGroup) {
```